src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ForeachToFor
+{
+    internal abstract class AbstractForeachToForCodeRefactoringProvider
+        : CodeRefactoringProvider
+    {
+        private const string get_Count = nameof(get_Count);
+        private const string get_Item = nameof(get_Item);
+
+        private const string Length = nameof(Length);
+        private const string Count = nameof(Count);
+
+        private static readonly ImmutableArray<string> s_KnownInterfaceNames =
+            new string[] { typeof(IList<>).FullName, typeof(IReadOnlyList<>).FullName, typeof(IList).FullName }.ToImmutableArray();
+
+        protected abstract SyntaxNode GetForeachStatement(SyntaxToken token);
```

Comment:
this type should have type parameters so we can strongly type the inputs/outputs of these methods when we implement them in C#/VB.  See most of our other refactorings/analyzers.

Reply:
will see. since if we add support ForEachVariableStatement it can't be fixed to ForEachStatement.

Reply:
Base type can be parameterized on both ForEachStatmentSyntax and ForEachVariableStatement.  Or, you can just parameterize it on CommonForEachStatementSyntax.

Reply:
+1 for using CommonForEachStatementSyntax